server/utils/plugins.js

```diff
@@ -0,0 +1,38 @@
+const assert = require('bsert');
+const Config = require('bcfg');
+
+function getPluginEndpoints(config) {
+  assert(config instanceof Config, 'Expected a bcfg object');
+  const beforeCoreMiddleware = [];
+  const afterCoreMiddleware = [];
+  // load list of plugins and local plugins from config
+  const plugins = config.array('plugins', []);
+  plugins.push(...config.array('localPlugins', []));
+
+  // go through each plugin id in the list
+  for (let plugin of plugins) {
+    assert(
+      typeof plugin === 'string',
+      `Expected plugin name to be a string instead got a ${typeof plugin}`
+    );
+
+    // require the `server` entrypoint for each (skip if no entry)
+    let module;
+    try {
+      module = require(`${plugin}/server`);
+    } catch (e) {}
+
+    if (!module) continue;
+
+    const { configs, endpoints = [] } = module;
+
+    if (configs && configs.afterCoreMiddleware)
```

Comment:
How is `configs` defined?

Reply:
Just an object with different config properties for now. Seemed useful to have just in case, but not necessary either since it checks if `configs` exists first at all. Once we have an idea of how the backend plugins can be useful and should be structured, would definitely be nice to document. I'm open to another name though. Configs seemed better than metadata for the backend plugins, but don't want it to be confused with the bcfg configs that are added elsewhere.

Reply:
Just an object with different config properties for now. Seemed useful to have just in case, but not necessary either since it checks if `configs` exists first at all. Once we have an idea of how the backend plugins can be useful and should be structured, would definitely be nice to document. I'm open to another name though. Configs seemed better than metadata for the backend plugins, but don't want it to be confused with the bcfg configs that are added elsewhere.

Reply:
Just trying to be thoughtful as to not have too many different places to add configuration options. We have `~/.bpanel/secrets.json`, which can apply to any plugin right? Then `[plugin]/server/configs`. If a user wants to configure the plugin after it is installed via npm, would they have to go into `node_modules/[plugin]/server/configs` to change the configuration?

Reply:
This would not be for user controlled configs but rather for the developer to dictate how their plugin will interact with bPanel. `secrets.env` and `bpanelConfigs` (set w/ `bcfg` either through env vars, command line args, or in the `config.js`) are user set configurations. Note for example that `afterCoreMiddleware` or `beforeCoreMiddleware` is  a config that a plugin dev will want to dictate given the functionality of their plugin (for example the faucet probably wants to insist their middleware comes before the wallet endpoints) and not something that would be updated by the user. This is the same as the frontend `metadata`, things like `version`, `icon`, `path`, etc. As you note, this is cumbersome for a user to get to, so is basically dev-level configs only, i.e.  a way for the plugin dev to tell bPanel a little bit about itself. An example of user defined configurations would be [these in the faucet](https://github.com/bpanel-org/faucet#development-and-setup).

I definitely do plan to document this once the API is hardened and the docs site is updated :)

Reply:
I don't feel like having to go into `node_modules` to set up configuration really makes sense, its not a pattern that i've ever seen. The first idea that comes to mind is a merge, where there is a `~/.bpanel/server/configs.json` (open to other file names) that would be plugin name -> config object, and the `[plugin]/server/configs` gets merged like so:

`plugin_config = Object.assign(node_modules/[plugin]/server/configs, ~/.bpanel/server/configs/[plugin])`

This allows for devs to have default configs and then people that want to deploy an instance of bpanel just need a file to reconfigure however they want